contrib/submit-protocol/src/App/MarketplaceForm.tsx

```diff
@@ -0,0 +1,198 @@
+];
+
+export default class MarketplaceForm extends React.Component<IMarketplaceProps, IMarketplaceState> {
+  public static defaultProps: Partial<IMarketplaceProps> = {
+    defaultURI: "https://api.github.com/repos/Microsoft/pai/contents/marketplace-v2",
+    defaultURIType: "GitHub" as MarketplaceUriType,
+    disabled: false,
+  };
+
+  public state = {
+    uri: this.props.defaultURI,
+    uriType: this.props.defaultURIType,
+    protocolOptions: [... defaultProtocolOptions],
+    uriConfigCallout: false,
+  };
+
+  private uriConfigCalloutBtn = React.createRef<HTMLDivElement>();
+
+  public componentDidMount() {
+    this.setProtocolOptions(null as any);
+  }
+
+  public render() {
+    return (
+      <Fabric>
+        <Stack gap={5} horizontal={true} verticalAlign="center">
+          <Stack>
+            <Dropdown
+              className={styles.dropdown}
+              placeholder="Select a protocol config file"
+              options={this.state.protocolOptions}
+              onChange={this.selectProtocol}
+              disabled={this.props.disabled}
+            />
+          </Stack>
+          <Stack maxWidth="5px">
+            <div ref={this.uriConfigCalloutBtn}>
+              <DefaultButton
+                onClick={this.toggleConfigCallout}
+                iconProps={{iconName: "ConfigurationSolid"}}
+                text="URI"
+                disabled={this.props.disabled}
+              />
+            </div>
+            <Callout
+              role="alertdialog"
+              target={this.uriConfigCalloutBtn.current}
+              onDismiss={this.closeConfigCallout}
+              setInitialFocus={true}
+              hidden={!this.state.uriConfigCallout}
+            >
+              <Stack padding={20}>
+                <TextField
+                  className={styles.textfiled}
+                  label="Marketplace URI"
+                  prefix={this.state.uriType || undefined}
+                  value={this.state.uri}
+                  onChange={this.setMarketplaceURI}
+                  onBlur={this.setProtocolOptions}
+                />
+              </Stack>
+            </Callout>
+          </Stack>
+        </Stack>
+      </Fabric>
+    );
+  }
+
+  private toggleConfigCallout = () => {
+    this.setState({uriConfigCallout: !this.state.uriConfigCallout});
+  }
+
+  private closeConfigCallout = () => {
+    this.setState({uriConfigCallout: false});
+  }
+
+  private setMarketplaceURI = (event: React.FormEvent<HTMLElement>, uri?: string) => {
+    if (uri !== undefined) {
+      this.setState({
+        uri,
+        uriType: "GitHub",
+      });
+    }
+  }
+
+  private setProtocolOptions = async (event: React.FocusEvent<HTMLInputElement>) => {
+    const protocolList = await this.getProtocolList(this.state.uri, this.state.uriType);
+    if (protocolList) {
+      const protocolOptions: IDropdownOption[] = [... defaultProtocolOptions];
```

Comment:
It's better to use defaultProtocolOptions.concat(protocolItem.map) that save a clone operation and also keep defaultProtocolOptions immutable.